frb_codegen/src/lib.rs

```diff
@@ -40,10 +40,19 @@ mod parser;
 mod source_graph;
 mod target;
 mod transformer;
-mod utils;
+pub mod utils;
 use error::*;
 
-pub fn frb_codegen(config: &config::Opts, all_symbols: &[String]) -> anyhow::Result<()> {
+pub fn frb_codegen(
+    configs: &[config::Opts],
+    index: usize,
```

Comment:
Hmm shall we put `index` here? Or is it possible to be inside `configs` etc

Reply:
We know there is a field `block_index` in `Opts`,  so there is already a field in charge of the index. And the reason why using  ` index: usize,` instead of `index: BlockIndex`(or just omit the parameter) is that I want to double-check the order of the current config is consistent with  that in `all_symbols` in `main.rs`, therefore  here comes  the double-check code:
```rust
pub fn frb_codegen(
    configs: &[config::Opts],
    index: usize,
    all_symbols: &[String],
) -> anyhow::Result<()> {
    let config = configs
        .iter()
        .find(|each| each.block_index == BlockIndex(index))
        .unwrap();
```
If the `unwrap()` doesn't make panic, that means it is safe to use `config.block_index`  afterwards for this config routine.

Reply:
Hmm... The API looks a bit less than ideal IMHO :/

Reply:
>  a bit less than ideal 
what do you want for this function signature? What is missing?

Reply:
Hopefully we can avoid passing such `index`. e.g. Is it possible we just pass in `config: config::Opts` etc

Reply:
> Is it possible we just pass in config: config::Opts

IMHO, no,  the configs are essential to catch global info. The global info is in used in `generate_dummy`. And the global info would be especially useful when doing [the shared task.](https://github.com/fzyzcjy/flutter_rust_bridge/issues/987#issuecomment-1407940083)

Reply:
I will revisit this a bit later when having a bit of time. This may need a bit of thinking

Reply:
Revisiting this problem. The new code

```
pub fn frb_codegen(
    configs: &[config::Opts],
    index: usize,
    all_symbols: &[String],
) -> anyhow::Result<()> {
    let config = configs
        .iter()
        .find(|each| each.block_index == BlockIndex(index))
        .unwrap();
```

looks equivalent to the old code https://github.com/fzyzcjy/flutter_rust_bridge/blob/master/frb_codegen/src/lib.rs

```
pub fn frb_codegen(config: &config::Opts, all_symbols: &[String]) -> anyhow::Result<()> {
```

especially given that, the `index` is never used in other places.

Thus, is it possible to keep the old function signature?

In addition, if we change signature it is a breaking change for everyone

Reply:
Alternatively what about this:

```
// an extra function, whoever wants multi-file uses this instead
pub fn frb_codegen_multi(configs: &[config::Opts], all_symbols) {
  for index in 0..something {
    frb_codegen(configs.iter().find(|each| each.block_index == BlockIndex(index)).unwrap(), all_symbols)?;
  }
}
```

Reply:
> frb_codegen_multi

Taking this one, I guess you mean defining `frb_codegen_multi(..)` in `lib.rs` right? In the case, it is easy to see `frb_codegen(..)` is enough for `pure_dart`,and only `frb_codegen_multi(..)` is enough for `pure_dart_multi`. 

But what the one in `frb_codegen/src/main.rs`? Since it is not know whether there is single or multi-blocks, I guess using the  `frb_codegen_multi(..)`  one?

In addition, a severe problem at present, since the `dummy` issue needs to be dealt with in the 1st config block, how do I catch the global info from `frb_codegen(...)` following the style?

Reply:
>In addition, if we change signature it is a breaking change for everyone

By the way, why would you think this signature change is a greate change? As far as I know, there are only 3 places using the signature `frb_codegen(..)`, and inside the new defined `frb_codegen(..)`, after :
```rust
  let config = configs
        .iter()
        .find(|each| each.block_index == BlockIndex(index))
        .unwrap();
```
Everything would be the same, except that you can get global info by `configs` instead of the current `config` when dealing each  block.

Reply:
> But what about the one in frb_codegen/src/main.rs? Since it is not known whether there is single or multi-blocks, I guess using the frb_codegen_multi(..) one?

Seems yes

> In addition, a severe problem at present, since the dummy https://github.com/fzyzcjy/flutter_rust_bridge/pull/1024#discussion_r1109133848 needs to be dealt within the 1st config block, how do I catch the global info from frb_codegen(...) following the style?

Hmm that's a problem

> By the way, why would you think this signature change is a greate change? As far as I know, there are only 3 places using the signature frb_codegen(..)

Because a lot of users of this library may be using this API - it is a *public* API.

Reply:
> (Maybe this style is enough if all shared stuff could be done in another shared config, but not at present.)

Oh, I am wrong. Even there is another auto-generated shared config intance, it is still essential to catch global information when dealing each regular block. For example, the path for shared block, the names of shared structs....

Reply:
I am busy with my code so let me write down a summarization to reduce brain context switch:

Problem: Suppose we want to generate api_1, api_2, api_3 three parts. Then, how to design the API?
Requirements: (1) Neat API (2) Best to be backward-compatible.

Reply:
>Because a lot of users of this library may be using this API - it is a public API.
oh, my god, I didn't see this. I thought the signature change this time would be as smooth as I did when introducing multi-blocks.

By the way, without changing signature, I've though introducing global constant variable. But it seems to be not feasible here, because the global instance(say `CONFIGS`) would be used in 
`lib.rs`, but it is not feasible before `frb_codegen` is called.

Reply:
What about this:

```rust
/// Need to add a line of comment, telling readers `all_configs` is used because need global information
pub fn frb_codegen_part_of_multi(
    config: &config::Opts,
    // add `all_` so it is less confusing. together with comments should be clearer to readers
    all_configs: &[config::Opts],
) {
 // the real main code
}

// this API is (1) a convenient shortcut for non-multi users (2) make it backward compatible
pub fn frb_codegen(config) {
  // just delegate to frb_codegen_part_of_multi
  frb_codegen_part_of_multi(config, vec![config])
}
```

Reply:
Now, I am looking forward to the feature of default/varadic paramter, If rust has this feature, there would be no public API that needs to be changed.

And for now, is it feasible to let users to use a macro say `frb_codegen!(...)`, so that it would be easy to use either `frb_codegen!(a,b)`,  `frb_codegen!(a,b,c)` and so on.  With this macro, it would be easily to do some break change inside without changing the signature.

Reply:
I guess https://github.com/fzyzcjy/flutter_rust_bridge/pull/1024#discussion_r1109183538 will not introduce any breaking change, and at the same time, allows to get global information (all_configs) easily

Reply:
> I guess [#1024 (comment)](https://github.com/fzyzcjy/flutter_rust_bridge/pull/1024#discussion_r1109183538) will not introduce any breaking change, and at the same time, allows to get global information (all_configs) easily

If I understand correctly, you are suggesting to refractor inside `frb_codegen(...)`?   If this is the case, where does the `all_configs` come from?

Reply:
Aside from using macro, there is another way without changing the signature. That is, new a field in `[Opt](config::Opts)` to contain essential global info, but this way just makes code too coupled.

Reply:
Suppose old code on master branch is:

```
pub fn frb_codegen(config) {
 [the-old-200-lines-of-code]
}
```

or, similarly, the code at this time in this PR is:

```
pub fn frb_codegen(configs, index) {
 [the-old-200-lines-of-code]
}
```

the new code is:

```
/// Need to add a line of comment, telling readers `all_configs` is used because need global information
pub fn frb_codegen_part_of_multi(
    config: &config::Opts,
    // add `all_` so it is less confusing. together with comments should be clearer to readers
    all_configs: &[config::Opts],
) {
 [the-old-200-lines-of-code] // JUST SIMPLY MOVE IT TO HERE
}

// this API is (1) a convenient shortcut for non-multi users (2) make it backward compatible
pub fn frb_codegen(config) {
  // just delegate to frb_codegen_part_of_multi
  frb_codegen_part_of_multi(config, vec![config])
}
```

Reply:
I see. So keeping the same code in `build.rs` for `pure_dart` and `main.rs` for `frb_code_gen`, modify `build.rs`  for `pure_dart_multi` with `frb_codegen_part_of_multi(...)` defined in `lib.rs` of `frb_code_gen`.

Reply:
Fixed in https://github.com/fzyzcjy/flutter_rust_bridge/pull/1024/commits/80edccbb1a12c6c21a5de2c78be772536335f670.